core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```diff
@@ -0,0 +1,1870 @@
+    case BIP32DerivationPathKeyId.byte => BIP32DerivationPathKeyId
+    case _: Byte                       => UnknownKeyId
+  }
+
+  final case object RedeemScriptKeyId extends PSBTOutputKeyId {
+    override val byte: Byte = 0x00.byteValue
+    type RecordType = OutputPSBTRecord.RedeemScript
+  }
+
+  final case object WitnessScriptKeyId extends PSBTOutputKeyId {
+    override val byte: Byte = 0x01.byteValue
+    type RecordType = OutputPSBTRecord.WitnessScript
+  }
+
+  final case object BIP32DerivationPathKeyId extends PSBTOutputKeyId {
+    override val byte: Byte = 0x02.byteValue
+    type RecordType = OutputPSBTRecord.BIP32DerivationPath
+  }
+
+  final case object UnknownKeyId extends PSBTOutputKeyId {
+    override val byte: Byte = Byte.MaxValue
+    type RecordType = OutputPSBTRecord.Unknown
+  }
+}
+
+sealed trait PSBTMap[+RecordType <: PSBTRecord] extends NetworkElement {
+  require(elements.map(_.key).groupBy(identity).values.forall(_.length == 1),
+          "All keys must be unique.")
+
+  def elements: Vector[RecordType]
+
+  def bytes: ByteVector =
+    elements
+      .sortBy(_.key)
+      .foldLeft(ByteVector.empty)(_ ++ _.bytes) :+ PSBTMap.separatorByte
+
+  protected def getRecords[KeyIdType <: PSBTKeyId](
+      key: KeyIdType,
+      keyIdFactory: PSBTKeyIdFactory[KeyIdType]): Vector[key.RecordType] = {
+    elements
+      .filter(element => keyIdFactory.fromByte(element.key.head) == key)
+      .asInstanceOf[Vector[key.RecordType]]
+  }
+}
+
+object PSBTMap {
+  final val separatorByte: Byte = 0x00.byteValue
+}
+
+sealed trait PSBTMapFactory[
+    RecordType <: PSBTRecord, MapType <: PSBTMap[RecordType]]
+    extends Factory[MapType] {
+  def recordFactory: Factory[RecordType]
+
+  def constructMap(elements: Vector[RecordType]): MapType
+
+  lazy val empty: MapType = constructMap(Vector.empty)
+
+  override def fromBytes(bytes: ByteVector): MapType = {
+    @tailrec
+    def loop(
+        remainingBytes: ByteVector,
+        accum: Vector[RecordType]): Vector[RecordType] = {
+      if (remainingBytes.head == PSBTMap.separatorByte) {
+        accum
+      } else {
+        val record = recordFactory.fromBytes(remainingBytes)
+        val next = remainingBytes.drop(record.bytes.size)
+
+        loop(next, accum :+ record)
+      }
+    }
+
+    constructMap(loop(bytes, Vector.empty))
+  }
+}
+
+case class GlobalPSBTMap(elements: Vector[GlobalPSBTRecord])
+    extends PSBTMap[GlobalPSBTRecord] {
+  import org.bitcoins.core.psbt.GlobalPSBTRecord._
+  import org.bitcoins.core.psbt.PSBTGlobalKeyId._
+  require(getRecords(UnsignedTransactionKeyId).nonEmpty,
+          "A GlobalPSBTMap must have a Unsigned Transaction")
+
+  def unsignedTransaction: UnsignedTransaction = {
+    getRecords(UnsignedTransactionKeyId).head
+  }
+
+  def extendedPublicKeys: Vector[XPubKey] = {
+    getRecords(XPubKeyKeyId)
+  }
+
+  def version: Version = {
+    getRecords(VersionKeyId).headOption.getOrElse(Version(UInt32.zero))
```

Comment:
any reason why we are defaulting to `UInt32.zero` here?

Reply:
https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#version-numbers

If no version is given then it is 0